PWGJE/Core/JetFinder.cxx

```diff
@@ -14,82 +14,49 @@
-}
 
-/// Sets the background subtraction estimater pointer
-void JetFinder::setBkgE()
-{
-  if (bkgSubMode == BkgSubMode::rhoAreaSub || bkgSubMode == BkgSubMode::constSub) {
-    bkgE = decltype(bkgE)(new fastjet::JetMedianBackgroundEstimator(selRho, jetDefBkg, areaDefBkg));
-  } else {
-    if (bkgSubMode != BkgSubMode::none) {
-      LOGF(error, "requested subtraction mode not implemented!");
-    }
-  }
-}
-
-/// Sets the background subtraction pointer
-void JetFinder::setSub()
-{
-  //if rho < 1e-6 it is set to 1e-6 in AliPhysics
-  if (bkgSubMode == BkgSubMode::rhoAreaSub) {
-    sub = decltype(sub){new fastjet::Subtractor{bkgE.get()}};
-  } else if (bkgSubMode == BkgSubMode::constSub) { //event or jetwise
-    constituentSub = decltype(constituentSub){new fastjet::contrib::ConstituentSubtractor{bkgE.get()}};
-    constituentSub->set_distance_type(fastjet::contrib::ConstituentSubtractor::deltaR);
-    constituentSub->set_max_distance(constSubRMax);
-    constituentSub->set_alpha(constSubAlpha);
-    constituentSub->set_ghost_area(ghostArea);
-    constituentSub->set_max_eta(bkgEtaMax);
-    constituentSub->set_background_estimator(bkgE.get()); //what about rho_m
-  } else {
-    if (bkgSubMode != BkgSubMode::none) {
-      LOGF(error, "requested subtraction mode not implemented!");
-    }
-  }
+  mSubUtils = std::make_unique<JetBkgSubUtils>(jetR, bkgPhiMin, bkgPhiMax, bkgEtaMin, bkgEtaMax, constSubAlpha, constSubRMax, ghostAreaSpec);
+  mSubUtils->setJetAlgorithmAndScheme(algorithmBkg, recombSchemeBkg);
 }
 
 /// Performs jet finding
 /// \note the input particle and jet lists are passed by reference
 /// \param inputParticles vector of input particles/tracks
 /// \param jets veector of jets to be filled
 /// \return ClusterSequenceArea object needed to access constituents
-fastjet::ClusterSequenceArea JetFinder::findJets(std::vector<fastjet::PseudoJet>& inputParticles, std::vector<fastjet::PseudoJet>& jets) //ideally find a way of passing the cluster sequence as a reeference
+fastjet::ClusterSequenceArea JetFinder::findJets(std::vector<fastjet::PseudoJet>& inputParticles, std::vector<fastjet::PseudoJet>& jets) // ideally find a way of passing the cluster sequence as a reeference
 {
   setParams();
-  setBkgE();
   jets.clear();
 
-  if (bkgE) {
-    bkgE->set_particles(inputParticles);
-    setSub();
-  }
-  if (constituentSub) {
-    inputParticles = constituentSub->subtract_event(inputParticles);
-  }
+  sub = mSubUtils->setSub(inputParticles, mRho, bkgSubMode);
+
   fastjet::ClusterSequenceArea clusterSeq(inputParticles, jetDef, areaDef);
-  jets = sub ? (*sub)(clusterSeq.inclusive_jets()) : clusterSeq.inclusive_jets();
+
+  if (bkgSubMode == BkgSubMode::rhoPerpConeSub) {
+    sub = mSubUtils->setSub(inputParticles, mRho, bkgSubMode, clusterSeq.inclusive_jets());
+  }
+
+  jets = (mRho > DBL_EPSILON && bkgSubMode != BkgSubMode::constSub) ? (sub)(clusterSeq.inclusive_jets()) : clusterSeq.inclusive_jets();
```

Comment:
what is DBL_EPSILON?

Reply:
DBL_EPSILON is the0 for floating number since mRho is float. So what I am doing here is that if Rho=0 (which means bkgsub is not set) then don't do any subtraction. The problem here is that sub is no longer a pointer in which we can just say (!sub), but now it's just a normal object. Any idea on how to do it better than this ?